File: w.js
const axios = require('axios');
const cheerio = require('cheerio');
const fetch = require('node-fetch');
const fs = require('fs');

// Telegram Bot configuration
const TELEGRAM_BOT_TOKEN = '7857441004:AAEZnfkRs-uSFSSjMzdmDHH1cJpTflVioio';
const CHANNEL_ID = '@news_SL1';
const BASE_URL = `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}`;

// Variable to store the last sent news ID
let lastSentNewsId = null;

// Fetch the latest news
async function fetchLatestNews() {
    const url = 'https://esana.com.lk/c/read';

    try {
        const response = await axios.get(url);
        if (response.status !== 200) {
            console.log(`Failed to fetch the page. Status code: ${response.status}`);
            return [null, null, null];
        }

        const $ = cheerio.load(response.data);
        const newsBlocks = $('div.post-block');
        let latestNews = null;
        let latestId = -1;

        newsBlocks.each((index, element) => {
            const link = $(element).find('a').attr('href');
            if (!link) return;

            const newsId = parseInt(link.split('/').pop(), 10);
            if (isNaN(newsId)) return;

            const title = $(element).find('h3.axil-post-title').text().trim() || 'No Title';
            const imageUrl = $(element).find('img').attr('src') || null;
            const timeText = $(element).find('div.post-metas a').text().trim() || 'Unknown time';
            const fullLink = `https://esana.com.lk/${link}`;

            if (newsId > latestId) {
                latestId = newsId;
                latestNews = { title, message: `Latest News: ${title}\nPosted: ${timeText}\nLink: ${fullLink}`, imageUrl };
            }
        });

        return latestNews ? [latestNews, latestId] : [null, null];
    } catch (error) {
        console.error('Error fetching the page:', error);
        return [null, null, null];
    }
}

// Send message to Telegram with optional image
async function sendToTelegram(message, imageUrl = null) {
    try {
        if (imageUrl) {
            // Download the image
            const imageResponse = await fetch(imageUrl);
            const imageBuffer = await imageResponse.buffer();

            const form = new FormData();
            form.append('chat_id', CHANNEL_ID);
            form.append('caption', message);
            form.append('parse_mode', 'Markdown');
            form.append('photo', imageBuffer, 'image.jpg');

            const telegramApiUrl = `${BASE_URL}/sendPhoto`;
            const response = await axios.post(telegramApiUrl, form, { headers: form.getHeaders() });
            console.log(`Response: ${response.status}, ${response.data}`);
            return response.status === 200;
        } else {
            // Send only text if no image URL is provided
            const telegramApiUrl = `${BASE_URL}/sendMessage`;
            const response = await axios.post(telegramApiUrl, {
                chat_id: CHANNEL_ID,
                text: message,
                parse_mode: 'Markdown',
            });
            return response.status === 200;
        }
    } catch (error) {
        console.error('Error sending to Telegram:', error);
        return false;
    }
}

// Monitor for new news every 30 seconds
setInterval(async () => {
    const [latestNews, latestId] = await fetchLatestNews();
    if (latestNews && latestId && latestId !== lastSentNewsId) {
        const { title, message, imageUrl } = latestNews;
        if (await sendToTelegram(message, imageUrl)) {
            console.log(`Sent to Telegram: ${message}`);
        } else {
            console.log("Failed to send to Telegram.");
        }
        lastSentNewsId = latestId; // Update the last sent news ID
    } else {
        console.log("No new news to send.");
    }
}, 30000);  // Wait 30 seconds before checking again

File: w.py
import requests
from bs4 import BeautifulSoup
import time
import os

# Telegram Bot configuration
TELEGRAM_BOT_TOKEN = '7857441004:AAEZnfkRs-uSFSSjMzdmDHH1cJpTflVioio'
CHANNEL_ID = '@news_SL1'
BASE_URL = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}"

# Variable to store the last sent news ID
last_sent_news_id = None

def fetch_latest_news():
    url = 'https://esana.com.lk/c/read'
    response = requests.get(url)
    if response.status_code != 200:
        print(f'Failed to fetch the page. Status code: {response.status_code}')
        return None, None, None

    soup = BeautifulSoup(response.text, 'html.parser')
    news_blocks = soup.find_all('div', class_='post-block')
    latest_news = None
    latest_id = -1  # To track the highest ID

    for news in news_blocks:
        # Extract link and find the numeric ID
        link_tag = news.find('a', href=True)
        link = f"https://esana.com.lk/{link_tag['href']}" if link_tag else None
        if not link:
            continue
        
        # Extract the numeric part of the URL
        try:
            news_id = int(link.split('/')[-1])  # Gets the number at the end of the URL, e.g., 104870
        except ValueError:
            continue  # Skip if the ID is not a number

        # Extract news title
        title_tag = news.find('h3', class_='axil-post-title')
        title = title_tag.get_text(strip=True) if title_tag else "No Title"

        # Extract image URL
        image_tag = news.find('img')
        image_url = image_tag['src'] if image_tag else None

        # Extract time
        time_tag = news.find('div', class_='post-metas').find('a')
        time_text = time_tag.get_text(strip=True) if time_tag else "Unknown time"

        # Check if this is the latest news based on the ID
        if news_id > latest_id:
            latest_id = news_id
            latest_news = (title, f"Latest News: {title}\nPosted: {time_text}\nLink: {link}", image_url)

    return latest_news if latest_news else (None, None, None), latest_id

def send_to_telegram(message, image_url=None):
    """Send message to the Telegram channel with an image as the main post and text as caption."""
    if image_url:
        # Send the image with the message as a caption
        try:
            telegram_api_url = f"{BASE_URL}/sendPhoto"
            payload = {
                'chat_id': CHANNEL_ID,
                'caption': message,
                'parse_mode': 'Markdown'
            }
            files = {'photo': requests.get(image_url).content}
            response = requests.post(telegram_api_url, data=payload, files=files)
            print(f"Response: {response.status_code}, {response.text}")
            return response.ok
        except Exception as e:
            print(f"Failed to send image to Telegram. Error: {e}")
            return False
    else:
        # Send only text if no image URL is provided
        telegram_api_url = f"{BASE_URL}/sendMessage"
        payload = {
            'chat_id': CHANNEL_ID,
            'text': message,
            'parse_mode': 'Markdown'
        }
        response = requests.post(telegram_api_url, json=payload)
        return response.ok

# Monitor for new news every 5 minutes
while True:
    result, latest_id = fetch_latest_news()
    
    if result is None:
        print("No news found or failed to fetch news.")
    else:
        title, news_message, image_url = result
        if latest_id and latest_id != last_sent_news_id:  # Check if this news ID has already been sent
            if send_to_telegram(news_message, image_url):
                print(f"Sent to Telegram: {news_message}")
            else:
                print("Failed to send to Telegram.")
            last_sent_news_id = latest_id  # Update the last sent news ID
        else:
            print("No new news to send.")
    time.sleep(30)  # Wait 30S before checking again
